### PR TITLE
More flexible preparation limits

### PR DESCRIPTION
**Pull request by andrewstart**
* Combines the shared code of WebGLPrepare and CanvasPrepare into one BasePrepare class.
* Makes the rate limiting functionality more configurable by changing it to be class based and adding a class for limiting based on time spent preparing.
* Adds hooks for both WebGL and Canvas for preparing PIXI.Text and PIXI.TextStyle objects.
* Adds unit tests for the rate limiting and basic functionality of BasePrepare.

I did some testing of GPU upload times on a mobile device and found that texture upload time was quite variable and generally expensive, so I wanted to be able to limit preparation to a specified amount of time to attempt to preserve the framerate instead of loading as quickly as possible.
I also found that calculating the font properties was generally worse than uploading any single texture to the GPU, so added the hooks to prepare that stuff ahead of time using the same system.

### CODE REVIEW

**Comment by ivanpopelyshev** [comment]
I think we also need PIXELS per frame. 2048x2048 and that kind off stuff

**Comment by andrewstart** [comment]
That's a little more complex than I wanted to get into. It also can't be a primary limiting metric, as it is only valid for textures.

**Comment by ivanpopelyshev** [comment]
@themoonrat there's something about fonts, can you check it?

File: src/deprecation.js [hunk 1]
```diff
@@ -642,3 +656,49 @@ Object.defineProperty(core.utils, '_saidHello', {
...
+    set()
+    {
+        warn('PIXI.CanvasPrepare.UPLOADS_PER_FRAME has been removed. Please set '
+            + 'renderer.plugins.prepare.limiter.maxItemsPerFrame on your renderer');
+        // because we don't have a reference to the renderer, we can't actually set
+        // the uploads per frame, so we'll have to stick with the warning.
+    },
+    get()
+    {
+        return 4;
```
Comment (englercj) [inline src/deprecation.js]: Same as above

**Comment by andrewstart** [comment]
Deprecations have been fixed.

**Pull request merged**